### ISSUE TITLE
`cast` not respected when used on a decorated class

### ISSUE THREAD

**Issue by Viicos** [microsoft/pyright#8506]
I am using a class decorator that is incorrectly typed as returning a different class, even though this is not the case at runtime (the example above is simplified but in reality we are facing limitations which would require HK type vars).

To overcome this issue, I'm using `cast` to lie to the type checker about the true type. However, `pyright` seems to be confused:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAIYDOlApngPrwI1kDGVMAUN6wDZWVCALk5QxUBAO6cAJjWBQ5rABQBrftSGxENANpE4AXQCUUALQA%2Bbc32GtAOkfSAAks58BUAEIjxEqZwCdDCMOspeZEx6BCZiAMRQAPIA0twARlAAvFDslDDhZF7KxsbcIDQAbjTkvKHMymmmUAkAKjpQYAoARGldmIJdBF1klAAWYACuvDJQaTRQXV5dAPyB1MF1NA2FTQkAcmBQdOAgy1Ato0iCrGAoMCDkMkisMIIwo-MI5RVIk4IABt9qrUov8cjVeJwgA)

```python
from typing import Any, assert_type, cast


class A:
    pass


def dec(klass: type[Any]) -> type[A]: ...


@dec
class B:
    pass

assert_type(B, type[A])  # OK

b = cast(B, B())

reveal_type(b)  # Type of "b" is "A", should be "B"?
assert_type(b, B)  # No error? This contradicts the previous `reveal_type` call

```

**Comment by Viicos** [comment]
Ah yes makes sense! When using `cast(B, B())`, pyright resolves `B` from the previous (decorated) definition, hence it still infers it a `A`.
